flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java

```diff
@@ -420,6 +420,7 @@ private void checkAsyncException() throws IOException {
         if (e != null) {
 
             asyncProducerException = null;
+            numRecordsOutErrorsCounter.inc();
```

Comment:
Is there a test that covers this? Looks like we missed this as well in the initial PR.

Reply:
I improved the `testErrorPropagationAndErrorsCounter` to explicitly test this addition

Reply:
Actually, this looks like an incorrect redundant increment. It seems like it is already being incremented once in the mailbox executor (which I think is the right place to do the increment, not here).

I believe the `testErrorPropagationAndErrorsCounter` is not catching this incorrect dual increment because in that test you're bypassing the callbacks to directly inject the async exception. This is probably a hint that the `@VisibleForTesting setAsyncException()` method is a bad test injection.

Reply:
On that matter, for leaner test cases additionally I think we should:

1. Break up `testErrorPropagationAndErrorsCounter` test into 2 separate tests that individually test error propagation and the error counter, separately.
2. consider removing the `testNumRecordsOutErrorsCounterMetric` test as it seems to have overlapping coverage with your new test (or improve that one). The `testNumRecordsOutErrorsCounterMetric` is really bloated, though, as it's using an awkward way just to trigger an async exception through the callback.

Reply:
The logic for counter is duplicated because it is possible for write/flush/WriterCallback-mailbox to throw the exception. When the exception is thrown, the counter should also be incremented. 

Ideally, we can increment the counter once when I detect the exception (inside the WriterCallback, before the mailbox) but we have no control over the Counter and it is not volatile.

I also am questioning the value of such a counter because the exception should cause a DAG restart.

Reply:
Agreed about the test overlap, but I think I need to test propagation through the 3 paths, in which counter increment and exception throwing make sense to couple?